File: app/assets/javascripts/services/EntityService.js
/*
 * Copyright (C) 2016 Language Technology Group and Interactive Graphics Systems Group, Technische Universität Darmstadt, Germany
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

define([
    'angular'
], function (angular) {
    'use strict';
    angular.module('myApp.entityservice', ['play.routing'])
        .factory('EntityService', ['$rootScope', 'playRoutes', '$timeout', function ($rootScope, playRoutes, $timeout) {
            var keywordScope = null;
            var entityScope = null;

            var toggleEntityGraph = true;
            var toggleKeywordGraph = true;

            return {

                subscribeBlacklist: function(scope, callback) {
                    var handler = $rootScope.$on('notifying-service-event', callback);
                    scope.$on('$destroy', handler);
                },

                subscribeBlacklistKeyword: function(scope, callback) {
                    var handler = $rootScope.$on('notifying-service-event2', callback);
                    scope.$on('$destroy', handler);
                },

                blacklist: function(ids) {
                    // TODO: Move to EntityController
                    playRoutes.controllers.NetworkController.blacklistEntitiesById(ids).get().then(function(response) {
                        $rootScope.$emit('notifying-service-event', { parameter: ids, response: response });
                    });
                },

                blacklistKeyword: function (keyword, id) {
                    playRoutes.controllers.KeywordNetworkController.blacklistKeywordByKeyTerm(keyword).get().then(function (response) {
                        $rootScope.$emit('notifying-service-event2', {parameter: [id], response: response});
                    });
                },

                whitelist: function(entity, type, docId, entId = "empty") {
                    playRoutes.controllers.EntityController.whitelistEntity(
                      entity.text,
                      entity.start,
                      entity.end,
                      type,
                      docId,
                      entId
                    ).get().then(function(response) {
                        $rootScope.$emit('notifying-service-event', { parameter: entity, response: response });
                    });
                },

                toggleTags: function (state, scope) {
                    playRoutes.controllers.KeywordNetworkController.toggleTags(state).get().then(function () {
                        scope.reloadGraph();
                    })
                },

                reloadKeywordGraph: function (toggle = false) {
                    if(keywordScope != null){
                        keywordScope.checkTags(toggle);
                    }
                },

                reloadEntityGraph: function () {
                    if(entityScope){
                      entityScope.reloadGraph();
                    }
                },

                setKeywordScope: function (scope) {
                    keywordScope = scope;
                },

                setEntityScope: function (scope) {
                    entityScope = scope;
                },

                highlightKeywords: function (keywords) {
                    if(keywordScope != null){
                        keywordScope.highlightKeywordNodes(keywords);
                    }
                },

                removeKeywordNodeHighlight: function () {
                    if(keywordScope != null) {
                        keywordScope.removeKeywordNodeHighlight();
                    }
                },

                highlightEntities: function (entities) {
                    if(entityScope != null){
                        entityScope.highlightEntityNodes(entities);
                    }
                },

                removeEntityNodeHighlight: function () {
                    if(entityScope != null) {
                        entityScope.removeEntityNodeHighlight();
                    }
                },

                setToggleEntityGraph: function (state) {
                    toggleEntityGraph = state;
                    if(keywordScope && entityScope){
                        if(toggleEntityGraph){
                            setTimeout(function() {
                                entityScope.resizeNetwork();
                            }, 500);
                        }
                        else {
                            $timeout(function() {
                                keywordScope.resizeNetwork();
                            }, 500);
                        }
                    }
                    else {
                        console.log('scopes not defined');
                    }
                },

                toggleEntityGraph: function () {
                    toggleEntityGraph = ! toggleEntityGraph;
                    if(keywordScope && entityScope){
                        if(toggleEntityGraph){
                            setTimeout(function() {
                                entityScope.resizeNetwork();
                            }, 500);
                        }
                        else {
                            $timeout(function() {
                                keywordScope.resizeNetwork();
                            }, 500);
                        }
                    }
                    else {
                        console.log('scopes not defined');
                    }
                },

                getToggleEntityGraph: function () {
                    return toggleEntityGraph;
                },

                setToggleKeywordGraph: function (state) {
                    toggleKeywordGraph = state;
                    if(keywordScope && entityScope){
                        if(toggleKeywordGraph) {
                            $timeout(function() {
                                keywordScope.resizeNetwork();
                            }, 500);
                        }
                        else {
                            setTimeout(function() {
                                entityScope.resizeNetwork();
                            }, 500);
                        }
                    }
                    else {
                        console.log('scopes not defined');
                    }
                },

                toggleKeywordGraph: function () {
                    toggleKeywordGraph = ! toggleKeywordGraph;
                    if(keywordScope && entityScope){
                        if(toggleKeywordGraph){
                            $timeout(function() {
                                keywordScope.resizeNetwork();
                            }, 500);
                        }
                        else {
                            setTimeout(function() {
                                entityScope.resizeNetwork();
                            }, 500);
                        }
                    }
                    else {
                        console.log('scopes not defined');
                    }
                },

                getToggleKeywordGraph: function () {
                    return toggleKeywordGraph;
                },

                getTagsSelected() {
                    return keywordScope.getTagsSelected();
                },

                getHostAddress() {
                    playRoutes.controllers.KeywordNetworkController.getHostAddress().get().then(function (response) {
                        return response;
                    });
                }
            };
        }])
});

File: app/models/EntityService.scala
/*
 * Copyright (C) 2016 Language Technology Group and Interactive Graphics Systems Group, Technische Universität Darmstadt, Germany
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package models

import com.google.inject.ImplementedBy
import model.Entity
import model.EntityType.withName

case class Fragment(start: Int, end: Int)

@ImplementedBy(classOf[DBEntityService])
trait EntityService {

  def getByIds(ids: List[Long])(index: String): List[Entity]

  def blacklist(ids: List[Long])(index: String): Boolean
  def undoBlacklist(ids: List[Long])(index: String): Boolean
  def getBlacklisted()(index: String): List[Entity]

  def merge(focalId: Long, duplicates: List[Long])(index: String): Boolean
  def undoMerge(focalIds: List[Long])(index: String): Boolean
  def getMerged()(index: String): Map[Entity, List[Entity]]

  def changeName(id: Long, newName: String)(index: String): Boolean
  def changeType(id: Long, newType: String)(index: String): Boolean

  def getEntityFragments(docId: Long)(index: String): List[(Entity, Fragment)]

  def getTypes()(index: String): List[String]
}

class DBEntityService extends EntityService {

  override def getByIds(ids: List[Long])(index: String): List[Entity] = {
    Entity.fromDBName(index).getByIds(ids)
  }

  override def blacklist(ids: List[Long])(index: String): Boolean = {
    val entityAPI = Entity.fromDBName(index)
    ids.map(entityAPI.delete).forall(identity)
  }

  override def undoBlacklist(ids: List[Long])(index: String): Boolean = {
    val entityAPI = Entity.fromDBName(index)
    ids.map(entityAPI.undoDelete).forall(identity)
  }

  override def getBlacklisted()(index: String): List[Entity] = {
    Entity.fromDBName(index).getBlacklisted()
  }

  override def merge(focalId: Long, duplicates: List[Long])(index: String): Boolean = {
    Entity.fromDBName(index).merge(focalId, duplicates)
  }

  override def undoMerge(focalIds: List[Long])(index: String): Boolean = {
    val entityAPI = Entity.fromDBName(index)
    focalIds.map(entityAPI.undoMerge).forall(identity)
  }

  override def getMerged()(index: String): Map[Entity, List[Entity]] = {
    Entity.fromDBName(index).getDuplicates()
  }

  override def changeName(id: Long, newName: String)(index: String): Boolean = {
    Entity.fromDBName(index).changeName(id, newName)
  }

  override def changeType(id: Long, newType: String)(index: String): Boolean = {
    Entity.fromDBName(index).changeType(id, withName(newType))
  }

  override def getEntityFragments(docId: Long)(index: String): List[(Entity, Fragment)] = {
    Entity.fromDBName(index).getEntityDocumentOffsets(docId).map { case (e, start, end) => (e, Fragment(start, end)) }
  }

  override def getTypes()(index: String): List[String] = {
    Entity.fromDBName(index).getTypes().map(_.toString)
  }
}
